### PR TITLE
Introduce verbose mode to traditional make build

### PR DESCRIPTION
**Pull request by gkostka**
It is nice when compilation log is short and clear. End user could
also decide how to call make:
 - make VERBOSE=1 to see full compilation log
 - make to see short log

### CODE REVIEW

**Comment by gkostka** [comment]
Yep, You are right. Should i fix it or You will dd it after merge ?

**Comment by gkostka** [comment]
Done

**Comment by gkostka** [comment]
I made some improvements:
 - refactor SILENT to Q
 - "\r\t" after echo (output alignment)
 - linux kernel like output
 - silent mode for scripts

File: Makefile [hunk 1]
```diff
@@ -174,38 +182,47 @@ $(OBJECTS): $(OUTPUT)include/distortos/distortosConfiguration.h
...
 
 .PHONY: size
 size:
-	$(SIZE) -B $(filter %.elf,$(^))
+	$(Q)$(SIZE) -B $(filter %.elf,$(^))
 
 .PHONY: clean
 clean:
-	$(RM) -r $(OUTPUT)
+	$(Q)$(RM) -r $(OUTPUT)
```
Comment (KamilSzczygiel) [inline Makefile]: I think this particular command should not be silenced, because otherwise there will be completely no output for "make clean"...

**Comment by gkostka** [comment]
Done. Squashed to one commit :). Removed Q from clean target.

File: Makefile [hunk 2]
```diff
@@ -174,38 +182,47 @@ $(OBJECTS): $(OUTPUT)include/distortos/distortosConfiguration.h
...
 
 .PHONY: size
 size:
-	$(SIZE) -B $(filter %.elf,$(^))
+	$(Q)$(SIZE) -B $(filter %.elf,$(^))
 
 .PHONY: clean
 clean:
-	$(RM) -r $(OUTPUT)
+	(RM) -r $(OUTPUT)
```
Comment (KamilSzczygiel) [inline Makefile]: a little bit too much (;

Comment (gkostka) [inline Makefile]: Ok. Fixed.

2015-10-12 16:20 GMT+02:00 Kamil Szczygieł <notifications@github.com>:

**Comment by KamilSzczygiel** [comment]
Funny thing - neither Eclipse nor my shell (Bash in Arch Linux) likes the "\r\t" - it just prints it the way it is written in the Makefile... Does that work like it should on your system? I guess that we will have to stay with regular spaces for that.

Maybe texts like "BIN", "HEX", "DMP" and "LSS" would be better than "OBJDUMP" and "OBJCOPY"? I don't have a strong opinion on that, so how do you think?

**Comment by KamilSzczygiel** [comment]
Doing "help echo" suggests that you need to do sth like that:

    echo -e "whatever\r\tyou\r\tlike"

Without "-e" the escapes are not interpreted. If this works on your system, then I think the easiest option is to use spaces, as this would work reliably on any system (probably <: ).

**Comment by gkostka** [comment]
Done. Tabs & CR changed to spaces. "OBJDUMP" and "OBJCOPY" changet to
"BIN", "HEX", "DMP" and "LSS".

2015-10-12 16:27 GMT+02:00 Kamil Szczygieł <notifications@github.com>:

**Comment by KamilSzczygiel** [comment]
Merged manually with minor formatting changes. Thanks!

BTW - was the echo " CXX\r\t" working fine on your system (without the "-e" flag)?

**Comment by gkostka** [comment]
BTW - was the echo " CXX\r\t" working fine on your system (without the "-e"
flag)?

Yes. It was working fine only in console. In eclipse compilation log was
not so awesome, so spaces solution should be most portable.

2015-10-12 18:38 GMT+02:00 Kamil Szczygieł <notifications@github.com>:

**Comment by KamilSzczygiel** [comment]
I see... Spaces are OK too, just not so fancy as "\r\t" (; Now it would be quite nice to hide the "echo ..." lines in verbose mode. I guess some clever macro should be used, as the other solution (ifeq ... endif) doesn't look so well.

It is worth noting that in Eclipse short compilation log shouldn't be used - at least in my opinion. Properly configured build output parser needs full invocations of the compiler to work fine. With short compilation log, Eclipse's indexer will not work at all. http://stackoverflow.com/a/26094959/157344

**Comment by jasmin-j** [comment]
> I see... Spaces are OK too, just not so fancy as "\r\t" (;

echo -e "..." should work on every system. I am not sure for MSYS, but it works with the Unix tools
on Windows, which I use in the company. In Linux it will work on every system.

> Now it would be quite nice to hide the "echo ..." lines in verbose mode. I guess some clever macro
> should be used, as the other solution (ifeq ... endif) doesn't look so well.

You did this already ;)

> It is worth noting that in Eclipse short compilation log shouldn't be used - at least in my opinion.

Suggestion 1:
If you change the default behaviour to verbose logging, then we could use the "-s" make command
option to toggle to short logging. With this approach nobody will need to add "VERBOSE=1" to the
Eclipse build command, as it is the default. "make -s" is the preferred way to silence make!
It is possible to check the "-s" command line option within the makefile.

Suggestion 2:
Because you use already a macro to print, I would change all the "$(call PRETTY_PRINT,xxxxx)"
lines to a generic print macro, which gets the text as argument. Doing so, you have only one point
where you can change the real used command.
Changing from spaces to \t is then done at one position only.
If you like, I can add it together with my other makefile changes.

**Comment by KamilSzczygiel** [comment]
On Monday 12 of October 2015 14:24:33 Jasmin Jessich wrote:
> echo -e "..." should work on every system. I am not sure for MSYS, but it
> works with the Unix tools on Windows, which I use in the company. In Linux
> it will work on every system.

The problem with "\r\t" is that Eclipse insists on interpreting "\r" as a new 
line, so the output looks really bad (Eclipse's console is pretty dumb). 
That's why I think spaces are preferred here.

If we would change the behaviour to the opposite one (default - verbose), then 
this particular issue would be much smaller - verbose log will look good in 
Eclipse (or any other IDE) and if the user enables short log then it's his/her 
problem (;

> > It is worth noting that in Eclipse short compilation log shouldn't be used
> > - at least in my opinion.
> Suggestion 1:
> If you change the default behaviour to verbose logging, then we could use
> the "-s" make command option to toggle to short logging. With this approach
> nobody will need to add "VERBOSE=1" to the Eclipse build command, as it is
> the default. "make -s" is the preferred way to silence make! It is possible
> to check the "-s" command line option within the makefile.

It's hard to decide here... The approach with "make VERBOSE=1" is good that it 
is consistent with the thing I implemented in tup build system. In tup you 
pass "--verbose" to print everything, so it's the opposite to "make -s". This 
way the systems would not be so consistent (no flag == short log)...

To be really honest, I actually don't like the short log and don't use it <: 
It's useful only if you have a package and want to build it - in normal 
development cycle I always prefer to see everything.
 
> Suggestion 2:
> Because you use already a macro to print, I would change all the "$(call
> PRETTY_PRINT,xxxxx)" lines to a generic print macro, which gets the text as
> argument. Doing so, you have only one point where you can change the real
> used command.

What exactly do you mean here? Do you think about passing two words (in two 
arguments) to the macro and printing that nicely in the macro (adding tabs, 
spaces or whatever)? I was thinking about that too today, but the approach 
with \t without using \r has a small drawback that if the first word is longer 
than one tab, then the columns are no longer aligned so pretty and there's not 
much we can do about that.

We could also just drop the nice column-alignment and do it in the most simple 
way - with single space.

> Changing from spaces to \t is then done at one position only.
> If you like, I can add it together with my other makefile changes.

As for your changes - the smaller the change will be, the easier it is to 
incorporate. And I mean both "small commits" and "small pull requests" - for 
example one controversial thing from the pull request with 10 commits (let's 
say that it's the last one) can block the whole series, even though this 
particular change has no connection to the other ones.

As for the format of pretty printing - I don't think we should waste much time 
on that... Yes, this may look pretty cool, but in the end users of distortos 
want an RTOS, not a pretty compilation output (;

**Comment by gkostka** [comment]
The problem with "\r\t" is that Eclipse insists on interpreting "\r" as a
new
line, so the output looks really bad (Eclipse's console is pretty dumb).
That's why I think spaces are preferred here.

We could use just "\t" instead of "\r\t". It might be messy on terminals
with "\t" equals to 4 spaces.


As for the format of pretty printing - I don't think we should waste much
time
on that... Yes, this may look pretty cool, but in the end users of
distortos
want an RTOS, not a pretty compilation output (;

In some cases (for example Jenkins) users want to see clear output without
whole compiler flags and billion lines of the same compiler options. It
might be a good idea to print compilation flags once in silent mode.

I have also noticed that CMake output in verbose mode has silent print &
compile line. Like that:
[ 46%] Building C object lwext4/CMakeFiles/lwext4.dir/ext4_xattr.c.o
cd /home/gko/devel/workspace/lwext4-git/build_generic/lwext4 &&
/usr/bin/cc  ......blablabla


2015-10-12 23:49 GMT+02:00 Kamil Szczygieł <notifications@github.com>:

**Comment by KamilSzczygiel** [comment]
On Monday 12 of October 2015 15:09:27 gkostka wrote:
> We could use just "\t" instead of "\r\t". It might be messy on terminals
> with "\t" equals to 4 spaces.

Let's stay with spaces. The column alignment looks nice, but I really have 
nothing against removing that and reverting to plain "CC<single-
space>filename.c".

Even now - when the PRETTY_PRINT macro is not very sophisticated and you have 
to insert the spaces "manually" it's not such a big problem, as there are less 
than 10 invocations and this number will not rise so dramatically in the near 
future.

I was wondering yesterday whether it could be automated, so that PRETTY_PRINT 
macro would receive 2 arguments and align them internally, but I don't think 
this can be done with make only - there are no functions to get the length of 
string or to replace only part of the string. Using anything more 
sophisticated (AWK or shell) seems like an overkill, as this is makes very 
little difference in the project...
 
> In some cases (for example Jenkins) users want to see clear output without
> whole compiler flags and billion lines of the same compiler options.

We already have that (; I'm only arguing against overcomplicating this to make 
it more pretty (;

> It
> might be a good idea to print compilation flags once in silent mode.

Cannot be done easily, as files in the tree have different flags - mostly 
include paths.

> I have also noticed that CMake output in verbose mode has silent print &
> compile line. Like that:
> [ 46%] Building C object lwext4/CMakeFiles/lwext4.dir/ext4_xattr.c.o
> cd /home/gko/devel/workspace/lwext4-git/build_generic/lwext4 &&
> /usr/bin/cc  ......blablabla

But does this line serve any purpose other than repeating the information that 
can be found in the next line, thus making the log twice as long? We could - 
in theory - add support for VERBOSE=2, where you'd see everything, but would 
that really be useful?

**Pull request closed without merging**